### ISSUE TITLE
some ideas

### ISSUE THREAD

**Issue by qmx** [conradkleinespel/rooster#3]
I hate issues for discussing stuff, but there we go:

I was trying to build the same thing you've done with rooster, with two big differences

- using [libsodium](https://download.libsodium.org/doc/) via [sodiumoxide](http://dnaq.github.io/sodiumoxide/sodiumoxide/index.html)

The library is ultra fast and provides encryption which is going to last. (It's taken from djb's nacl after all), and the main consequence for this one would be having a better key derivation scheme - we definitely don't want to have SHA256 for keys :)

- using sqlite3 as storage instead of a single flat file

my reasoning for this one is having a file format which is well documented, stable and with all the bonus points of being almost drop-in compatible with android - which means less hassle for doing the mobile app.

Thoughts? I'd be more than happy to discuss more on this, since I don't want to duplicate effort :)

**Comment by conradkleinespel** [comment]
Hi @qmx,

Thanks a lot for your comment !

As can be seen in the updated `README.md`, the `sha256` KDF will soon be replaced with something more solid like `scrypt`.

Other people have suggested using `sqlite3` too. And it does sound interesting. I like that we could have access to database level constraints. We'd still need to document the schema and specify migrations if the schema changes.

I plan to address these 2 issues (better KDF and file format) soon. Including a way to easily change the schema as Rooster evolves, without the need for the end user to know about that.

Feel free to post a link to your project. I'd be happy to see what's in there :smiley: 

Best regards,
Conrad

**Comment by qmx** [comment]
> I plan to address these 2 issues (better KDF and file format) soon. Including a way to easily change the schema as Rooster evolves, without the need for the end user to know about that.

For schema changes, one of the things that could be done was to actually expose the schema version in a plain JSON fashion, like this:

```json
{
  "version":1,
  "another plain text data":"like salt",
  "passwords": "AAAAAAAAAAA" //cyphertext
}
```

This would make it easier for us to actually migrate data from a schema to another, and would solve the first problem I'm facing trying to implement the key derivation part: since you're marshalling the data straight to the opaque file, I can't peek for the schema version without trying to unmarshall it and let it fail :)

Thoughts?

**Comment by conradkleinespel** [comment]
Yes. Having the file version available without the need to decrypt would be good indeed.

And that is the plan for the next file format. From there on, I was thinking of keeping compatibility with the current file format for like 2 months and then just drop it, so people can upgrade easily for 2 months. And after 2 months, we could document a 2-step upgrade path in the `README.md` and the Rooster website for people that haven't yet upgraded. They'll be asked to download an older version of Rooster followed by a newer one so as to upgrade from the current file format to the versioned one first, and then from the versioned one to the very latest one.

What do you think ?

**Comment by qmx** [comment]
> From there on, I was thinking of keeping compatibility with the current file format for like 2 months and then just drop it, so people can upgrade easily for 2 months.

This can be automated, can't it? I mean, we could just convert it on the first load and it will be immediately saved with the new format.. That way all data format upgrades can be transparent to users.

TL;DR if there's no version header exposed assume "the big encrypted blob"

**Comment by conradkleinespel** [comment]
Yes. It could be. And it will be.

I'd like to not keep this in the code forever if possible though. It should definitely be automated at the start, so that's what I'll start with.

**Comment by conradkleinespel** [comment]
Hello @qmx,

I just published a [new release of Rooster](https://github.com/conradkleinespel/rooster/releases/tag/v2.0.0), which uses `scrypt` for key derivation.

**If you want to install it, please make sure you backup your existing Rooster file first**. I tested as well as possible, but I do write bugs sometimes.

The upgrade path from v1 should be straightforward: install Rooster v2 and just start using it.

Best regards and a happy new year,
Conrad

**Comment by conradkleinespel** [comment]
@qmx I almost forgot: about using SQLite. It does indeed sound interesting. I didn't get it into this release. But I suppose we could easily add it to an upcoming release, given that Rooster now has some code that helps upgrading file versions.

What would be the benefits of using SQLite for Rooster ?

**Comment by conradkleinespel** [comment]
I'll mark this as closed but please feel free to respond and/or open a new issue regarding a migration to sqlite :smiley:

**Comment by qmx** [comment]
@conradkleinespel yay! going to check it soon
